### PR TITLE
Settings with the new data architecture

### LINKED ISSUE #131: As a user, I want to be able to change my settings, so that I can personalise my experience.

### LINKED ISSUE #132: As a user who values privacy, I want to be able to control my personal information and choose what data is shared with other players, so that I can maintain my privacy and security.

### LINKED ISSUE #154: Create a settings class
Related to #131

### LINKED ISSUE #156: Create a privacy setting for the user about their profile's visibility
Related to #132

### PR DESCRIPTION
**Pull request by Maeeen**
:+1:

This PR closes #154, closes #156, closes #157, closes #131, closes #132.

### CODE REVIEW

File: app/src/main/java/com/github/geohunt/app/ui/components/profile/ProfilePageViewModel.kt [hunk 1]
```diff
@@ -84,6 +90,18 @@ open class ProfilePageViewModel(
...
+                    when (profileVisibilityRepository.getProfileVisibility(uid).first()) {
+                        ProfileVisibility.PUBLIC -> {}
+                        ProfileVisibility.PRIVATE -> { _isPrivate.value = true; return@launch }
+                        ProfileVisibility.FOLLOWING_ONLY -> {
+                            val isPrivate = !followRepository.doesFollow(uid, authedUid).first()
+                            _isPrivate.value = isPrivate
+                            if (isPrivate) return@launch
+                        }
+                    }
+                }
```
Comment (BoyeGuillaume) [inline app/src/main/java/com/github/geohunt/app/ui/components/profile/ProfilePageViewModel.kt:103]: **Note** (for future reference): We should introduce rules within firebase to enforce this mechanism

Comment (BoyeGuillaume) [reply app/src/main/java/com/github/geohunt/app/ui/components/profile/ProfilePageViewModel.kt:103]: It is just for future reference don't change anything

File: app/src/main/java/com/github/geohunt/app/ui/settings/RadioItem.kt [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+fun <T> RadioItem(
+    title: String,
+    value: T,
+    isSelected: Boolean,
+    onSelect: (T) -> Any,
+    modifier: Modifier = Modifier,
+    icon: ImageVector? = null,
+    subtitle: String? = null,
+    disabled: Boolean = false
+) {
```
Comment (BoyeGuillaume) [inline app/src/main/java/com/github/geohunt/app/ui/settings/RadioItem.kt:21]: Jetpack compose don't have radio button ???? Kinda sad really

**Review by BoyeGuillaume — CHANGES_REQUESTED** [review]
Look nice, one or two small tweeks and we can call it a day. I would understand if you just want to send it considering the tweeks are minor (and the shear amount of time we already spent on sdp this week). As such MP me if you just want to merge it

File: app/src/main/java/com/github/geohunt/app/data/settings/AppSettingsSerializer.kt [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+        get() = AppSettings()
+
+    override suspend fun readFrom(input: InputStream): AppSettings {
+        return try {
+            Json.decodeFromString(
+                string = input.readBytes().decodeToString(),
+                deserializer = AppSettings.serializer()
+            )
+        } catch (e: SerializationException) {
+            Log.e("WARN", "Could not parse app settings.")
```
[the remark this answers is not in the document]
Comment (Maeeen) [reply app/src/main/java/com/github/geohunt/app/data/settings/AppSettingsSerializer.kt:27]: This makes the app still usable, so it's not an error.

File: app/src/main/java/com/github/geohunt/app/ui/components/profile/ProfilePageViewModel.kt (hunk 1 above)
Comment (Maeeen) [reply app/src/main/java/com/github/geohunt/app/ui/components/profile/ProfilePageViewModel.kt:103]: > Note (for future reference): We should introduce rules within firebase to enforce this mechanism

Back-end ? 🙃

File: app/src/main/java/com/github/geohunt/app/ui/settings/RadioItem.kt (hunk 2 above)
Comment (Maeeen) [reply app/src/main/java/com/github/geohunt/app/ui/settings/RadioItem.kt:21]: No no, it does has one. However, not a stylized radio button as above. :(

File: app/src/main/java/com/github/geohunt/app/data/settings/AppSetting.kt [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+import kotlinx.coroutines.launch
+
+/**
+ * This represents an app settings for the view models.
+ */
+data class AppSetting<T>(val flow: Flow<T>, val defaultValue: T, val setter: suspend (T) -> Unit) {
+    /**
+     * Creates a new Mutable State flow that will reflect every change of the flow.
+     */
+    fun toOneWayMutableStateFlow(coroutineScope: CoroutineScope): MutableStateFlow<T> {
```
[the remark this answers is not in the document]
Comment (Maeeen) [reply app/src/main/java/com/github/geohunt/app/data/settings/AppSetting.kt:15]: Indeed, good catch!

File: app/src/main/java/com/github/geohunt/app/data/repository/ProfileVisibilityRepository.kt [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+    database: FirebaseDatabase = FirebaseDatabase.getInstance(),
+    private val ioDispatcher: CoroutineDispatcher = Dispatchers.IO
+): ProfileVisibilityRepositoryInterface {
+    private val visibilities = database.getReference("profileVisibilities")
+
+    override suspend fun getProfileVisibility(uid: String): Flow<ProfileVisibility> {
+        return visibilities.child(uid)
+            .snapshots
+            .map { it.getValue(Int::class.java) ?: 0 }
+            .map { it.coerceIn(0, ProfileVisibility.values().size - 1) }
```
[the remark this answers is not in the document]
Comment (Maeeen) [reply app/src/main/java/com/github/geohunt/app/data/repository/ProfileVisibilityRepository.kt:23]: Let's make it resilient, no? It's not really that important, is it? Especially in such a context. Creating errors implies handling them. :/

**Review by BoyeGuillaume — APPROVED** [review]

File: app/src/main/java/com/github/geohunt/app/ui/settings/RadioItem.kt (hunk 2 above)
Comment (BoyeGuillaume) [reply app/src/main/java/com/github/geohunt/app/ui/settings/RadioItem.kt:21]: :'(

**Review by BoyeGuillaume — DISMISSED** [review]

**Review by BoyeGuillaume — APPROVED** [review]

**Review by Nicoslz — APPROVED** [review]
Great code! Let's go for the pizza party!

**Pull request merged**